### PR TITLE
Adding generic SqlToSlackOperator

### LINKED ISSUE #24243: Create `SqlToSlackOperator`
### Body

Currently we have `SnowflakeToSlackOperator` & `PrestoToSlackOperator`.
This task is to create a generic `SqlToSlackOperator` to be compatible with all hooks that has `get_pandas_df` function.
The new operator should be in `providers/slack/transfers/sql_to_slack.py`, once completed we can also deprecate `SnowflakeToSlackOperator` & `PrestoToSlackOperator`.

A good resource for reference can be `SqlToS3Operator` which the first part of the operator (running query and creating dataframe should be identical), the 2nd part of creating the slack message should be very similar to what the current operators do.

### Committer

- [X] I acknowledge that I am a maintainer/committer of the Apache Airflow project.

### PR DESCRIPTION
**Pull request by alexkruc**
This PR is adding a generic `SqlToSlackOperator`, meaning everything that uses `DbApiHook` and implements `get_pandas_df` will now be able to send the results to Slack.
As part of this PR, I've also deprecated the current `PrestoToSlackOperator` and `SnowflakeToSlackOperator` - they now inherit from `SqlToSlackOperator`.
 
closes: #24243
reference: existing `PrestoToSlackOperator` and `SnowflakeToSlackOperator`, and the generic `SqlToS3Operator`.

### CODE REVIEW

File: tests/system/providers/slack/example_sql_to_slack.py
Comment (eladkal) [inline tests/system/providers/slack/example_sql_to_slack.py]: Lets not use Presto names as it's not longer Presto example :)
```suggestion
SQL_TABLE = os.environ.get("SQL_TABLE", "test_table")
```

File: tests/system/providers/slack/example_sql_to_slack.py
Comment (eladkal) [inline tests/system/providers/slack/example_sql_to_slack.py]:
```suggestion
        sql=f"SELECT col FROM {SQL_TABLE}",
```

File: tests/providers/presto/transfers/test_presto_to_slack.py
Comment (eladkal) [inline tests/providers/presto/transfers/test_presto_to_slack.py:45]: This test file should be minimal now.. only the tests that we need to verify that `PrestoToSlackOperator` is successfully "converted" into the new operator. All the other tests logic should be in `test_sql_to_slack.py`
The reasoning for this is that once we remove `PrestoToSlackOperator` we will also remove it's test file and by doing so we don't want to lose any tests that are important.

File: tests/providers/snowflake/transfers/test_snowflake_to_slack.py
Comment (eladkal) [inline tests/providers/snowflake/transfers/test_snowflake_to_slack.py:25]: Same comment like on the presto test file

**Review by eladkal — CHANGES_REQUESTED** [review]

File: tests/providers/presto/transfers/test_presto_to_slack.py
Comment (alexkruc) [reply tests/providers/presto/transfers/test_presto_to_slack.py:45]: I left the test as is, but changed the mocking location.
now, the test is verifying that the message that is getting built using the operator is actually triggered from the generic `sql_to_slack.SlackWebhookHook`.
All the other tests like rendering different df names and such are in `SqlToSlackOperator` tests.. So I think that the tests are pretty minimal as is, and we can delete them and you won't lose any "important" tests, or you will miss some logic..
My guideline here was end to end test `PrestoToSlackOperator`, and make sure that the actual call is called from `SqlToSlackOperator`.. if we remove `PrestoToSlackOperator` eventually, I see no issue here, as e2e is not important anymore :)

Or am I mistaken here?

**Comment by potiuk** [comment]
Rebased to fix selective check problem from #24665 after it's merged.

**Comment by eladkal** [comment]
@alexkruc please check test failures

File: airflow/providers/presto/transfers/presto_to_slack.py
Comment (eladkal) [inline airflow/providers/presto/transfers/presto_to_slack.py]: can we just remove the `TYPE_CHECKING` then?

File: airflow/providers/presto/transfers/presto_to_slack.py
Comment (eladkal) [inline airflow/providers/presto/transfers/presto_to_slack.py]: this is a breaking change. Can we avoid it?
I prefer to leave the defaults of the old operators as is. In the new operator we can set new defaults as we like

File: docs/apache-airflow-providers-slack/operators/sql_to_slack.rst
Comment (eladkal) [inline docs/apache-airflow-providers-slack/operators/sql_to_slack.rst]:
```suggestion
channel.
```
We don't support sending to more than 1 channel do we?

File: tests/providers/presto/transfers/test_presto_to_slack.py
Comment (eladkal) [reply tests/providers/presto/transfers/test_presto_to_slack.py:45]: fair point!

**Comment by Taragolis** [comment]
Just an idea: https://github.com/apache/airflow/pull/24660#issuecomment-1166502250 make sending to slack more generic 

Might be some side effect and multiple inheritance required in some cases or not.

File: airflow/providers/slack/transfers/sql_to_slack.py
Comment (eladkal) [inline airflow/providers/slack/transfers/sql_to_slack.py:106]: If I'm reading this right you are mixing between kwargs for BaseOperator and also kwargs for the underlying hook (SowflakeHook, PrestoHook, etc..) I don't think this is right.
We should have a dedicated dict to pass the hook params
similar to what we have in `SqlSensor`

https://github.com/apache/airflow/blob/37ab8fc76b4d977914ed54fc392b14cdb478d9d1/airflow/sensors/sql.py#L67-L68

To give specific example:
```
SqlToSlackOperator(
    ...,
    retries=3,
    hook_params= {"warehouse": "my_warehouse"}
)
```
 
Did I miss something?

File: airflow/providers/presto/transfers/presto_to_slack.py
Comment (alexkruc) [reply airflow/providers/presto/transfers/presto_to_slack.py]: Umm, we can, but I'm not sure how should I tackle this..
The thing is, that using both `slack_conn_id` and `slack_webhook_token`, although supported by the `SlackWbhookHook`, can lead to misunderstandings.
I added a check for this in `SqlToSlackOperator`, that if both are provided, we go out with an AirflowException.
If I leave this as is, with the default initialization of the `slack_conn_id`, if a user provides me a `slack_webhook_token`, it will fail (because both are provided - once by the user, and once by the initialization).

To solve this, I can add a check in `SqlToSlackOperator` that checks if `slack_conn_id` is provided, and the connection actually exists, but again it may lead to a confusing flow, as if `slack_conn_id` is provided and exists, it will be the one used by the `SlackWebhookHook`..

I found it to be easier to provide either one, but idk.. WDYT? Should I add a check whether the connection actually exists? or do you might have some other creative suggestion? :)

**Comment by Taragolis** [comment]
@eladkal let's keep track here

> The generic operator will handle this - if implemented right there will be no need for individual classes in each provider, again if you have comments on this lets please take them to #24663

I would suggest more generic way, `SqlToSlackOperator` this generic only for DBApi.

If we have something that I mentioned before:  base class in **slack-provider** `BasePandasDataFrameToSlackOperator` without implementation `get_pandas_df`

It could help to build not only `SqlToSlackOperator` but also some user-classes like `AwsDataWranglerDynamoDBToSlack`, `LocalDataFrameToSlack`, etc.

**Comment by Taragolis** [comment]
Additional benefits you need to test implementation how actually works part with sending DataFrame in `slack-provider` (consumer), and how to produce this DataFrame in actual implementation, like in `SqlToSlackOperator`.

If any bugs appear in slack part, we need need to fix it `slack-provider`, if on producer part, only need to fix on producer.
Same work with new features.

Again, just an idea and I only talk about benefits without negative part of this idea.

**Comment by eladkal** [comment]
yes we can, but this is a further enhancement. If alex wants to handle this as well he is more than welcome :) 
personally I'm OK with leaving it to a followup PR. It's OK to make a change in a few smaller PR.. it makes the review easier

**Comment by Taragolis** [comment]
> yes we can, but this is a further enhancement. If alex wants to handle this as well he is more than welcome :)
personally I'm OK with leaving it to a followup PR. It's OK to make a change in a few smaller PR.. it makes the review easier

Totally agree that it should be further enhancement, otherwise it easily took ages to implements SqlToSlackOperator, and deprecate in Snowflake and Presto.

I will look on this PR more close on Monday. Probably I would suggest something that I've already found during SnowflakeToSlack enhancement.

File: docs/apache-airflow-providers-slack/operators/sql_to_slack.rst
Comment (Taragolis) [reply docs/apache-airflow-providers-slack/operators/sql_to_slack.rst]: Oh... 

Slack Incoming Webhook based on [Slack App](https://api.slack.com/messaging/webhooks) can't change channel,  in other word one webhook = one channel

Also exists legacy [Incoming Webhook](https://slack.com/apps/A0F7XDUAZ-incoming-webhooks?tab=more_info) based on Deprecated Slack Integration which can change channel.

The problem was it's unclear what exactly intend to use SlackWebhookHook. Actually it could use both of them, but I don't think we need to focus on legacy features.

File: airflow/providers/presto/transfers/presto_to_slack.py
Comment (Taragolis) [inline airflow/providers/presto/transfers/presto_to_slack.py]: `webhook_token` in Extra marked as Deprecated

https://github.com/apache/airflow/blob/b678dc25f6604ddf603e42500e3e5583b6663df6/airflow/providers/slack/hooks/slack_webhook.py#L103-L111

File: airflow/providers/presto/transfers/presto_to_slack.py
Comment (eladkal) [inline airflow/providers/presto/transfers/presto_to_slack.py]: you can use `self.log.info`

Comment (alexkruc) [reply airflow/providers/presto/transfers/presto_to_slack.py]: thanks, changed :)

File: airflow/providers/slack/transfers/sql_to_slack.py
Comment (alexkruc) [reply airflow/providers/slack/transfers/sql_to_slack.py:106]: added `sql_hook_params` - thanks for the tip! :)

File: airflow/providers/presto/transfers/presto_to_slack.py
Comment (alexkruc) [reply airflow/providers/presto/transfers/presto_to_slack.py]: added a check for the actual connection existence in `PrestoToSlackOperator`, and then passing the proper values to the `SqlToslackOperator`..
LMK if it's ok :)

File: airflow/providers/presto/transfers/presto_to_slack.py
Comment (alexkruc) [reply airflow/providers/presto/transfers/presto_to_slack.py]: removed :)

File: airflow/providers/slack/transfers/sql_to_slack.py
Comment (eladkal) [inline airflow/providers/slack/transfers/sql_to_slack.py]: @potiuk need your advise on this one.

The `hook_params` was added in https://github.com/apache/airflow/pull/18718 /  https://github.com/apache/airflow/pull/19849 which were released on `2.3.0` The Slack provider min version is 2.2.0 so I think this code line should have cause failures?

2.2.0:
https://github.com/apache/airflow/blob/2.2.0/airflow/models/connection.py#L292

2.3.0:
https://github.com/apache/airflow/blob/2.3.0/airflow/models/connection.py#L321

 I wonder how it make sense that the CI is green?

~@alexkruc in any case we don't need the `hook_params` here.. the goal is just to check if the hook has the needed function, the user params are not required for that.~

Comment (eladkal) [reply airflow/providers/slack/transfers/sql_to_slack.py]: @alexkruc I think we have two options here.
1. backport the `get_hook` function into the operator code (temporarily, we will remove it once the provider will be Airflow>=2.3 compatible)
2. Leave the `PrestoToSlackOperator` `SnowflakeToSlackOperator` code as is with just a deprecation warning, in that solution we should also check Airflow version when using the operator as it works only with 2.3 (you can check `from airflow.version import version` )

I think option 1 is prefered as otherwise it means only users of Airflow>=2.3 will be able to use this new operator.

Comment (potiuk) [reply airflow/providers/slack/transfers/sql_to_slack.py]: Yep. 1) is preferred.

Comment (eladkal) [reply airflow/providers/slack/transfers/sql_to_slack.py]: A fix to pre-commit to make sure we don't have this case is in https://github.com/apache/airflow/pull/24706
Thanks @potiuk

File: airflow/providers/slack/transfers/sql_to_slack.py
Comment (eladkal) [inline airflow/providers/slack/transfers/sql_to_slack.py]:
```suggestion
            # when "apache-airflow-providers-slack" will depend on Airflow >= 2.3.
```
That description is enough.

File: airflow/providers/slack/transfers/sql_to_slack.py
Comment (eladkal) [inline airflow/providers/slack/transfers/sql_to_slack.py]:
```suggestion
     when "apache-airflow-providers-slack" will depend on Airflow >= 2.3.
```

File: airflow/providers/slack/transfers/sql_to_slack.py
Comment (eladkal) [inline airflow/providers/slack/transfers/sql_to_slack.py]: Shouldn't we pass the `sql_hook_params` as well? this is the whole reason for backporing the function isn't it?

Comment (alexkruc) [reply airflow/providers/slack/transfers/sql_to_slack.py]: your'e right :) fixed it :)

File: airflow/providers/slack/transfers/sql_to_slack.py
Comment (eladkal) [reply airflow/providers/slack/transfers/sql_to_slack.py]: @potiuk is there a way to make the new pre commit ignore a specific row?
The pre-commit fails the PR but @alexkruc checks the user Airflow version and have different code to handle base on version.
We can use only the backported version if there is no other option

Comment (potiuk) [reply airflow/providers/slack/transfers/sql_to_slack.py]: Sure - we can modify the pre-commit and ignore lines that for example contains a comment `# ignore 2.2 comp check`. It's a python code that we can modify in this PR. leaving it as an exercise for the reader :D

Comment (alexkruc) [reply airflow/providers/slack/transfers/sql_to_slack.py]: Reader as in myself? ;)
I'll take a look, will love to get familiar a bit with the pre-commit flow :)
I'll update here if I'll get stuck or something.

I just thought that checking for versions and matching will be easier to read for future operator maintainers, and also it's easier to deprecate later (just remove the if block), I think that using the generic methods (as in `conn.get_hook(hook_params=....)`) as much as possible is the better way here :)

Comment (alexkruc) [reply airflow/providers/slack/transfers/sql_to_slack.py]: @potiuk @eladkal added the ignore lines for compatibility check fix, you can find it in this commit:
https://github.com/apache/airflow/pull/24663/commits/f3a29fb07754e6edaea94ff3e7f81817933958a1
It should run as part of this PR and we can check that it works e2e :) (although it worked locally)

Comment (potiuk) [reply airflow/providers/slack/transfers/sql_to_slack.py]: Cool! That's what I meant. One small NIT: since you made a "generic" exclusion - let's make it `# ignore airflow compat check`, We have some 2.2, some 2.3 and some 2.4 checks already foreseen to be there for quite some time.

Comment (potiuk) [reply airflow/providers/slack/transfers/sql_to_slack.py]: BTW. Pre-comits are SUPER easy :D

Comment (alexkruc) [reply airflow/providers/slack/transfers/sql_to_slack.py]: Great! thanks for the comment :) I changed it here - https://github.com/apache/airflow/pull/24663/commits/efd07a59159b4be94988b174b99af93557fad8ba

> BTW. Pre-comits are SUPER easy :D

I agree! and besides being easy, they are also kinda fun 😄  The thing is, that as a user, and as a young contributor, I'm not often exposed to this, so it's nice learning something new about the Airflow eco-system ;)

**Review by eladkal — APPROVED** [review]
LGTM

**Pull request merged**
